### PR TITLE
Allow restricting arguments as orderable or comparable in the function signature

### PR DESCRIPTION
**Pull request by duanmeng**
In Presto, input to min/max is restricted to orderable types. 
For example, MAP type is not orderable. Velox currently allows MAP inputs.

This PR adds two APIs in `FunctionSignatureBuilder`: `orderableTypeVariable`
and `comparableTypeVariable`. Use type's `orderable` and `comparable` flag
added in https://github.com/facebookincubator/velox/pull/6770 to do the check 
during type binding in `SignatureBinder.tryBind`.

Part of https://github.com/facebookincubator/velox/issues/6718

### CODE REVIEW

File: velox/expression/FunctionSignature.h [hunk 1]
```diff
@@ -58,6 +60,27 @@ class SignatureVariable {
     return knownTypesOnly_;
   }
 
+  void setKnownTypesOnly(bool knownTypesOnly) {
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.h:63]: Let's keep this class immutable and remove these setters.

File: velox/expression/FunctionSignature.h [hunk 2]
```diff
@@ -328,9 +399,55 @@ class AggregateFunctionSignatureBuilder {
 
   AggregateFunctionSignatureBuilder& knownTypeVariable(
       const std::string& name) {
-    addVariable(
-        variables_,
-        SignatureVariable(name, "", ParameterType::kTypeParameter, true));
+    if (variables_.find(name) == variables_.end()) {
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.h:402]: Looks like this logic is to allow calling both orderableTypeVariable and knownTypeVariable in the same signature. Do we need this? How about we don't allow if and if necessary add knownOrderableTypeVariable later?

File: velox/expression/FunctionSignature.cpp [hunk 3]
```diff
@@ -229,11 +229,15 @@ SignatureVariable::SignatureVariable(
...
     ParameterType type,
-    bool knownTypesOnly)
+    bool knownTypesOnly,
+    bool orderableTypesOnly,
+    bool comaprableTypesOnly)
     : name_{std::move(name)},
       constraint_(constraint.has_value() ? std::move(constraint.value()) : ""),
       type_{type},
-      knownTypesOnly_(knownTypesOnly) {
+      knownTypesOnly_(knownTypesOnly),
```
Comment (mbasmanova) [inline velox/expression/FunctionSignature.cpp:238]: Can we add a check that the new properties apply only when isTypeParameter() true?

File: velox/expression/tests/SignatureBinderTest.cpp [hunk 4]
```diff
@@ -378,6 +378,53 @@ TEST(SignatureBinderTest, knownOnly) {
   }
 }
 
+TEST(SignatureBinderTest, orderableComparable) {
+  auto signature = exec::FunctionSignatureBuilder()
+                       .orderableTypeVariable("T")
```
Comment (mbasmanova) [inline velox/expression/tests/SignatureBinderTest.cpp:383]: orderable implies comparable, hence, no point specifying both

**Review by mbasmanova — COMMENTED** [review]
@duanmeng Nice change. Looks good % some minor comments.

File: velox/expression/FunctionSignature.h (hunk 2 above)
Comment (duanmeng) [reply velox/expression/FunctionSignature.h:402]: It seems we do not need it for now. It is better to make them mutually exclusive and keep `SignatureVarible` immutable. I will update this PR according to these principles.

File: velox/expression/FunctionSignature.h (hunk 1 above)
Comment (duanmeng) [reply velox/expression/FunctionSignature.h]: Got it, fixed.

File: velox/expression/FunctionSignature.cpp (hunk 3 above)
Comment (duanmeng) [reply velox/expression/FunctionSignature.cpp:238]: Added.

File: velox/expression/tests/SignatureBinderTest.cpp (hunk 4 above)
Comment (duanmeng) [reply velox/expression/tests/SignatureBinderTest.cpp:383]: Removed

**Review by mbasmanova — APPROVED** [review]
Thanks.

File: velox/expression/FunctionSignature.h [hunk 5]
```diff
@@ -258,12 +274,13 @@ class FunctionSignatureBuilder {
...
-  FunctionSignatureBuilder& knownTypeVariable(const std::string& name) {
-    addVariable(
-        variables_,
-        SignatureVariable(name, "", ParameterType::kTypeParameter, true));
-    return *this;
-  }
+  FunctionSignatureBuilder& knownTypeVariable(const std::string& name);
+
+  /// Orderable implies comparable, this method would enable
+  /// comaprableTypesOnly_ too.
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/expression/FunctionSignature.h]: Fixed

File: velox/expression/FunctionSignature.h [hunk 6]
```diff
@@ -79,6 +93,8 @@ class SignatureVariable {
   // This property only applies to type variables and indicates if the type
   // can bind to unknown or not.
   bool knownTypesOnly_ = false;
+  bool orderableTypesOnly_ = false;
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/expression/FunctionSignature.h:96]: Updated

**Pull request closed without merging**